arch-riscv: adding support for local interrupts

robhau opened the pull request:
Besides the standard RISC-V interrupts software, timer, and external interrupt, the RISC-V specification also offers the possibility to implement local interrupts. With this patch, we contribute an extension of RiscvInterrupts that enables connecting interrupt sources to the local interrupt controller. We assigned the local interrupts to machine-level and gave them the highest priority. If two local interrupts are pending, there exception code will be the tie-breaker (higher ID > lower ID). 32 Bit systems only recognize the local interrupts 16 to 31, 64 Bit systems 16 to 63.

I was not able to reopen the pull request https://github.com/gem5/gem5/pull/811 after my force push. Therefore, i opened a new pull request again. I hope that it is not problematic 😅 .

Change-Id: Iff8d34e740b925dce351c0c6f54f4bd37a647e0c

robhau committed:
arch-riscv: adding support for local interrupts

Besides the standard RISC-V interrupts software, timer, and external
interrupt, the RISC-V specification also offers the possibility to
implement local interrupts. With this patch, we contribute an extension
of RiscvInterrupts that enables connecting interrupt sources to the
local interrupt controller. We assigned the local interrupts to
machine-level and gave them the highest priority. If two local
interrupts are pending, there exception code will be the tie-breaker
(higher ID > lower ID). 32 Bit systems only recognize the local
interrupts 16 to 31, 64 Bit systems 16 to 63.

Change-Id: Iff8d34e740b925dce351c0c6f54f4bd37a647e0c

robhau committed:
arch-riscv: add missing override specifier

Change-Id: I15593d275505c0da3f0edea17fcd28497726cae4

rogerchang23424 commented on src/arch/riscv/interrupts.hh:
```diff
@@ -137,83 +82,32 @@ class Interrupts : public BaseInterrupts
-    }
+    void unserialize(CheckpointIn &cp) override;
+
+    Port &getPort(const std::string &if_name, PortID idx);
```
missing override here

rogerchang23424 commented on src/arch/riscv/faults.hh:
```diff
@@ -93,6 +94,54 @@ enum ExceptionCode : uint64_t
     INT_EXT_USER = 8,
     INT_EXT_SUPER = 9,
     INT_EXT_MACHINE = 11,
+    INT_LOCAL_16 = 16,
```
I would like to start with 0 like `INT_LOCAL_0, INT_LOCAL_1 ...`

Ref: https://five-embeddev.com/quickref/interrupts.html

robhau replied:
Do you mean `INT_LOCAL_0` to `INT_LOCAL_47`?

I do not understand your reference. In https://five-embeddev.com/quickref/interrupts.html, the local interrupts also start with 16 (`platform_defined16, platform_defined17 ...`)

rogerchang23424 replied:
> Do you mean `INT_LOCAL_0` to `INT_LOCAL_47`?

Yes

> I do not understand your reference. In https://five-embeddev.com/quickref/interrupts.html, the local interrupts also start with 16 (`platform_defined16, platform_defined17 ...`)

It means the interrupt ID 16 - X is defined by platform. It comment `Optional platform defined interrupt source 0.` in the description field for the interrupt ID 16. So I think use `INT_LOCAL_0` will make it clear to know the interrupt source it from 0.

Another example for using local interrupt in SiFive E31:

 
![clint interrupt](https://github.com/gem5/gem5/assets/32214817/43a9b839-f335-42e5-91c7-49dd8024cfcd)

robhau replied:
I think it is better to keep `INT_LOCAL_16 ...` to highlight that, e.g., `INT_LOCAL_i` corresponds to the i-th bit in `mie` and `mip`.

robhau replied:
I forgot to refresh my browser before writing the comment above. Thank you for the examples. I will change this.

rogerchang23424 commented on src/arch/riscv/RiscvInterrupts.py:
```diff
@@ -27,10 +28,35 @@
     cxx_class = "gem5::RiscvISA::Interrupts"
     cxx_header = "arch/riscv/interrupts.hh"
+
+    local_interrupt_pins = VectorIntSinkPin("Pins for local interrupts")
```
Just curious. Why we need to add `intSinkPin` in the `RiscvInterrupts`.

Don't you think post interrupt in the model without local_interrupt_pins connection  
`tc->getCpuPtr()->postInterrupt(tc->threadId(), <int_local_num>, 0)` is good?

robhau replied:
I think it would be better to modularize this, as it is more flexible. If you, for example, want to connect your device to an other hart, you just have to do a simple change in the configuration switch. If you want to post the interrupt directly from the external device model, you would have to count your hardware threads in order to obtain the correct thread.

rogerchang23424 replied:
Ack

robhau committed:
arch-riscv: start local interrupt IDs with 0

Change-Id: I8a83e47a50ab20acbad1256ff7dcb99f1b22cb3d

rogerchang23424 commented on src/arch/riscv/regs/misc.hh:
```diff
@@ -1438,6 +1440,54 @@ USTATUS_MASKS[enums::Num_RiscvType][enums::Num_PrivilegeModeSet] = {
+const RegVal LOCAL_44_MASK = 1ULL << 44;
+const RegVal LOCAL_43_MASK = 1ULL << 43;
+const RegVal LOCAL_42_MASK = 1ULL << 42;
+const RegVal LOCAL_41_MASK = 1ULL << 41;
+const RegVal LOCAL_40_MASK = 1ULL << 40;
+const RegVal LOCAL_39_MASK = 1ULL << 39;
+const RegVal LOCAL_38_MASK = 1ULL << 38;
+const RegVal LOCAL_37_MASK = 1ULL << 37;
+const RegVal LOCAL_36_MASK = 1ULL << 36;
+const RegVal LOCAL_35_MASK = 1ULL << 35;
+const RegVal LOCAL_34_MASK = 1ULL << 34;
+const RegVal LOCAL_33_MASK = 1ULL << 33;
+const RegVal LOCAL_32_MASK = 1ULL << 32;
+const RegVal LOCAL_31_MASK = 1ULL << 31;
+const RegVal LOCAL_30_MASK = 1ULL << 30;
+const RegVal LOCAL_29_MASK = 1ULL << 29;
+const RegVal LOCAL_28_MASK = 1ULL << 28;
+const RegVal LOCAL_27_MASK = 1ULL << 27;
+const RegVal LOCAL_26_MASK = 1ULL << 26;
+const RegVal LOCAL_25_MASK = 1ULL << 25;
+const RegVal LOCAL_24_MASK = 1ULL << 24;
+const RegVal LOCAL_23_MASK = 1ULL << 23;
+const RegVal LOCAL_22_MASK = 1ULL << 22;
+const RegVal LOCAL_21_MASK = 1ULL << 21;
+const RegVal LOCAL_20_MASK = 1ULL << 20;
+const RegVal LOCAL_19_MASK = 1ULL << 19;
+const RegVal LOCAL_18_MASK = 1ULL << 18;
+const RegVal LOCAL_17_MASK = 1ULL << 17;
+const RegVal LOCAL_16_MASK = 1ULL << 16;
```
Please align the number `INT_LOCAL_0`, `INT_LOCAL_1` ...

robhau replied:
Thank you for your hint. I already replaced this individual bit masks for every local interrupt with `mask(63,16)` from `bitfield.hh`, but I forgot to delete the mentioned variables 😅 .

rogerchang23424 commented on src/arch/riscv/regs/misc.hh:
```diff
@@ -1438,6 +1440,54 @@ USTATUS_MASKS[enums::Num_RiscvType][enums::Num_PrivilegeModeSet] = {
     },
 };
 
+const RegVal LOCAL_63_MASK = 1ULL << 63;
+const RegVal LOCAL_62_MASK = 1ULL << 62;
+const RegVal LOCAL_61_MASK = 1ULL << 61;
+const RegVal LOCAL_60_MASK = 1ULL << 60;
+const RegVal LOCAL_59_MASK = 1ULL << 59;
+const RegVal LOCAL_58_MASK = 1ULL << 58;
+const RegVal LOCAL_57_MASK = 1ULL << 57;
+const RegVal LOCAL_56_MASK = 1ULL << 56;
+const RegVal LOCAL_55_MASK = 1ULL << 55;
+const RegVal LOCAL_54_MASK = 1ULL << 54;
+const RegVal LOCAL_53_MASK = 1ULL << 53;
+const RegVal LOCAL_52_MASK = 1ULL << 52;
+const RegVal LOCAL_51_MASK = 1ULL << 51;
+const RegVal LOCAL_50_MASK = 1ULL << 50;
+const RegVal LOCAL_49_MASK = 1ULL << 49;
+const RegVal LOCAL_48_MASK = 1ULL << 48;
+const RegVal LOCAL_47_MASK = 1ULL << 47;
+const RegVal LOCAL_46_MASK = 1ULL << 46;
+const RegVal LOCAL_45_MASK = 1ULL << 45;
+const RegVal LOCAL_44_MASK = 1ULL << 44;
+const RegVal LOCAL_43_MASK = 1ULL << 43;
+const RegVal LOCAL_42_MASK = 1ULL << 42;
+const RegVal LOCAL_41_MASK = 1ULL << 41;
```
ditto

robhau committed:
arch-riscv: delete bit masks for local interrupts

Change-Id: I9ce189a697522f27e9f66819b77b145ea8d5ae24

rogerchang23424 reviewed (approved):
LGTM

robhau committed:
arch-riscv: mask for local interrupts

Function mask(63,16) is replaced by variable LOCAL_MASK.
mask(63,16) is assigned to LOCAL_MASK.

Change-Id: Id2caaa932de36e05151332e1ce31af2f55c0ac86

robhau committed:
Merge branch 'develop' into RISCVLocalInterrupts

BobbyRBruce commented:
@robhau : Do you mind if we "Squash merge" this PR into the develop branch?

robhau commented:
@BobbyRBruce I do not mind ☺️

rogerchang23424 commented on src/arch/riscv/interrupts.hh:
```diff
@@ -137,83 +82,32 @@ class Interrupts : public BaseInterrupts
+    Port &getPort(const std::string &if_name, PortID idx) override;
+
+    void raiseInterruptPin(uint32_t num);
+    void lowerInterruptPin(uint32_t num) {};
```
I think the device cloud not clear the interrupt pending without handling the `lowerInterruptPin`, is right?

robhau replied:
Yes, the device can not clear the interrupt, it has to be done in software. If i understand interrupt pins correctly, two consecutive raise() calls on a `IntSourcePin` do not lead to two `raiseInterruptPin()` calls at the `IntSinkPin` device. Therefore, an interrupt source that does not support interrupt clearing is still enforced to hold the interrupt pin until the ISR is entered. With an empty `lowerInterruptPin()` function, the interrupt source just calls `raise()` followed by `lower()`.

What do you think about adding an additional parameter in the configuration script that contains the IDs of interrupts that can be cleared by `lower()`? Then, in `lowerInterruptPin()`, interrupts are cleared if `num` is in this list of IDs.

rogerchang23424 replied:
I have no idea why you need to raise interrupt pins twice. IMHO, the gem5 will check interrupts after each instructions is commit. So we can just call `raise()` once in the interrupt source and the interrupt pending will be set. For cleaning the interrupt pending, it should be the process of ISR execution.

Just curious the interrupt pending will be clear before jumping the ISR?

robhau replied:
Please correct me if I understood your question wrong:

At the moment, local interrupt pending bits can only be cleared by writing to `mip`.

```asm
csrrc zero, mip, (1 << 16)
```
To enable consecutive interrupts from devices with periodic interrupts, e.g., a sensor, `lowerInterruptPin()` has to be called before the next `raiseInterruptPin()`.

rogerchang23424 replied:
Ok, I will check the issue again after long Chinese New Year holiday.
